File: Euclid.Storage/Database/Data/Room/RoomData.cs

using FluentNHibernate.Mapping;

namespace Euclid.Storage.Database.Data
{
    public class RoomMapping : ClassMap<RoomData>
    {
        public RoomMapping()
        {
            Table("room");
            Id(x => x.Id, "id");
            Map(x => x.OwnerId, "owner_id");
            Map(x => x.Name, "name");
            Map(x => x.Description, "description");
            Map(x => x.CategoryId, "category_id").Generated.Insert();
            Map(x => x.UsersNow, "visitors_now").Generated.Insert();
            Map(x => x.UsersMax, "visitors_max").Generated.Insert();
            Map(x => x.AccessType, "access_type").Generated.Insert().CustomType<RoomStatus>();
            Map(x => x.Password, "password").Generated.Insert();
            Map(x => x.ModelId, "model_id");
            Map(x => x.CCTs, "ccts").Generated.Insert();
            Map(x => x.Wallpaper, "wallpaper").Generated.Insert();
            Map(x => x.Floor, "floor").Generated.Insert();
            Map(x => x.Rating, "rating").Generated.Insert();
            Map(x => x.ShowName, "show_name").Generated.Insert();
            Map(x => x.SuperUsers, "super_users").Generated.Insert();

            References(x => x.OwnerData, "owner_id")
                .ReadOnly()
                .Cascade.None()
                .NotFound.Ignore();

            References(x => x.Category, "category_id")
                .ReadOnly()
                .Cascade.None()
                .NotFound.Ignore();
        }
    }

    public class RoomData
    {
        public virtual int Id { get; set; }
        public virtual int OwnerId { get; set; }
        public virtual string Name { get; set; }
        public virtual string Description { get; set; }
        public virtual int CategoryId { get; set; }
        public virtual RoomStatus AccessType { get; set; }
        public virtual string Password { get; set; }
        public virtual int ModelId { get; set; }
        public virtual string CCTs { get; set; }
        public virtual int Wallpaper { get; set; }
        public virtual int Floor { get; set; }
        public virtual string Landscape { get; set; }
        public virtual PlayerData OwnerData { get; set; }
        public virtual NavigatorCategoryData Category { get; set; }
        public virtual int Rating { get; set; }
        public virtual bool IsPrivateRoom => OwnerId > 0;
        public virtual bool IsPublicRoom => OwnerId == 0;
        public virtual int UsersNow { get; set; }
        public virtual int UsersMax { get; set; }
        public virtual bool ShowName { get; set; }
        public virtual bool SuperUsers { get; set; }

        public static RoomStatus ToStatusEnum(int roomAccess)
        {
            switch (roomAccess)
            {
                case 1:
                    return RoomStatus.CLOSED;
                case 2:
                    return RoomStatus.PASSWORD;
                default:
                    return RoomStatus.OPEN;

            }
        }
    }

    public enum RoomStatus
    {
        OPEN = 0,
        CLOSED = 1,
        PASSWORD = 2
    }
}

File: euclid-storage/src/main/java/euclid/storage/database/data/RoomData.java
package euclid.storage.database.data;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;

@Entity
@Table(name = "room")
public class RoomData {

    @Id
    @Column(name = "id")
    private int id;

    @Column(name = "owner_id")
    private int ownerId;

    @Column(name = "name")
    private String name;

    @Column(name = "description")
    private String description;

    @Column(name = "category_id", insertable = false)
    private int categoryId;

    @Column(name = "visitors_now", insertable = false)
    private int usersNow;

    @Column(name = "visitors_max", insertable = false)
    private int usersMax;

    @Enumerated(EnumType.ORDINAL)
    @Column(name = "access_type", insertable = false)
    private RoomStatus accessType;

    @Column(name = "password", insertable = false)
    private String password;

    @Column(name = "model_id")
    private int modelId;

    @Column(name = "ccts", insertable = false)
    private String ccts;

    @Column(name = "wallpaper", insertable = false)
    private int wallpaper;

    @Column(name = "floor", insertable = false)
    private int floor;

    @Column(name = "rating", insertable = false)
    private int rating;

    @Column(name = "show_name", insertable = false)
    private boolean showName;

    @Column(name = "super_users", insertable = false)
    private boolean superUsers;

    @ManyToOne(fetch = FetchType.LAZY, optional = true)
    @JoinColumn(name = "owner_id", insertable = false, updatable = false)
    private PlayerData ownerData;

    @ManyToOne(fetch = FetchType.LAZY, optional = true)
    @JoinColumn(name = "category_id", insertable = false, updatable = false)
    private NavigatorCategoryData category;

    @Transient
    private String landscape;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getOwnerId() {
        return ownerId;
    }

    public void setOwnerId(int ownerId) {
        this.ownerId = ownerId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getCategoryId() {
        return categoryId;
    }

    public void setCategoryId(int categoryId) {
        this.categoryId = categoryId;
    }

    public int getUsersNow() {
        return usersNow;
    }

    public void setUsersNow(int usersNow) {
        this.usersNow = usersNow;
    }

    public int getUsersMax() {
        return usersMax;
    }

    public void setUsersMax(int usersMax) {
        this.usersMax = usersMax;
    }

    public RoomStatus getAccessType() {
        return accessType;
    }

    public void setAccessType(RoomStatus accessType) {
        this.accessType = accessType;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public int getModelId() {
        return modelId;
    }

    public void setModelId(int modelId) {
        this.modelId = modelId;
    }

    public String getCcts() {
        return ccts;
    }

    public void setCcts(String ccts) {
        this.ccts = ccts;
    }

    public int getWallpaper() {
        return wallpaper;
    }

    public void setWallpaper(int wallpaper) {
        this.wallpaper = wallpaper;
    }

    public int getFloor() {
        return floor;
    }

    public void setFloor(int floor) {
        this.floor = floor;
    }

    public int getRating() {
        return rating;
    }

    public void setRating(int rating) {
        this.rating = rating;
    }

    public boolean isShowName() {
        return showName;
    }

    public void setShowName(boolean showName) {
        this.showName = showName;
    }

    public boolean isSuperUsers() {
        return superUsers;
    }

    public void setSuperUsers(boolean superUsers) {
        this.superUsers = superUsers;
    }

    public PlayerData getOwnerData() {
        return ownerData;
    }

    public void setOwnerData(PlayerData ownerData) {
        this.ownerData = ownerData;
    }

    public NavigatorCategoryData getCategory() {
        return category;
    }

    public void setCategory(NavigatorCategoryData category) {
        this.category = category;
    }

    public String getLandscape() {
        return landscape;
    }

    public void setLandscape(String landscape) {
        this.landscape = landscape;
    }

    public boolean isPrivateRoom() {
        return ownerId > 0;
    }

    public boolean isPublicRoom() {
        return ownerId == 0;
    }

    public static RoomStatus toStatusEnum(int roomAccess) {
        return RoomStatus.toStatusEnum(roomAccess);
    }
}
